deployment/terraform/keycloak.go

```diff
@@ -0,0 +1,279 @@
+
+	if !t.config.ExternalAuthProviderSettings.DevelopmentMode {
+		command = "start"
+	}
+
+	sshc, err := extAgent.NewClient(t.output.KeycloakServer.PublicIP)
+	if err != nil {
+		return err
+	}
+
+	// Check if the keycloak database exists
+	result, _ := sshc.RunCommand(`sudo -iu postgres psql -l | grep keycloak 2> /dev/null`)
+	if len(result) == 0 {
+		mlog.Info("keycloak database not found, creating it and associated role")
+		_, err = sshc.RunCommand(`sudo -iu postgres psql <<EOSQL
+		CREATE USER keycloak WITH PASSWORD 'mmpass';
+		CREATE DATABASE keycloak OWNER keycloak;
+		GRANT ALL PRIVILEGES ON DATABASE keycloak TO keycloak;
+		EOSQL
+		'`)
+		if err != nil {
+			return fmt.Errorf("failed to setup keycloak database: %w", err)
+		}
+
+		if t.config.ExternalAuthProviderSettings.KeycloakDBDumpURI != "" && t.config.ExternalAuthProviderSettings.KeycloakRealmFilePath == "" {
+			err := t.IngestKeycloakDump()
+			if err != nil {
+				return fmt.Errorf("failed to ingest keycloak dump: %w", err)
+			}
+		}
+	}
+
+	// Check if we should use a custom dump, a custom realm file or the default one
+	if t.config.ExternalAuthProviderSettings.KeycloakRealmFilePath != "" {
+		_, err := sshc.UploadFile(t.config.ExternalAuthProviderSettings.KeycloakRealmFilePath, "/opt/keycloak/keycloak-"+t.config.ExternalAuthProviderSettings.KeycloakVersion+"/data/import/mattermost-realm.json", true)
+		if err != nil {
+			return fmt.Errorf("failed to upload keycloak realm file: %w", err)
+		}
+		extraArguments = append(extraArguments, "--import-realm")
+	} else if t.config.ExternalAuthProviderSettings.KeycloakDBDumpURI == "" {
+		mlog.Info("No realm file or database dump provided, using loadtest's default keycloak realm configuration")
+
+		keycloakRealmFile, err := assets.AssetString("mattermost-realm.json")
+		if err != nil {
+			return fmt.Errorf("failed to read default keycloak realm file: %w", err)
+		}
+
+		_, err = sshc.Upload(strings.NewReader(keycloakRealmFile), "/opt/keycloak/keycloak-"+t.config.ExternalAuthProviderSettings.KeycloakVersion+"/data/import/mattermost-realm.json", true)
+		if err != nil {
+			return fmt.Errorf("failed to upload default keycloak realm file: %w", err)
+		}
+		extraArguments = append(extraArguments, "--import-realm")
+	}
+
+	// Values for the keycloak.env file
+	keycloakEnvFileContents := []string{
+		// Enable health endpoints
+		"KC_HEALTH_ENABLED=true",
+		// Setup admin user
+		"KEYCLOAK_ADMIN=" + t.config.ExternalAuthProviderSettings.KeycloakAdminUser,
+		"KEYCLOAK_ADMIN_PASSWORD=" + t.config.ExternalAuthProviderSettings.KeycloakAdminPassword,
+		// Ensure Java JVM has enough memory for large imports
+		"JAVA_OPTS=-Xms1024m -Xmx2048m",
+		// Logging
+		"KC_LOG_FILE=/opt/keycloak/keycloak-" + t.config.ExternalAuthProviderSettings.KeycloakVersion + "/data/log/keycloak.log",
+		"KC_LOG_FILE_OUTPUT=json",
+		// Database
+		"KC_DB_POOL_MIN_SIZE=20",
+		"KC_DB_POOL_INITIAL_SIZE=20",
+		"KC_DB_POOL_MAX_SIZE=200",
```

Comment:
Question: would we need to modify this to support 100K users? If so, would it make sense to move it to config? (Out of scope for this PR obviously).

Reply:
Ok yeah that's a good point. I guess we will see in our big test.